Specify custom wallet directory with -walletdir param

meshcollider opened the pull request:
Closes #11348 

Adds a `-walletdir` parameter which specifies a directory to use for wallets, allowing them to be stored separately from the 'main' data directory. Creates a new `wallets/` directory in datadir if this is the first time running, and defaults to using it if it exists.

Includes tests and release notes. Things which might need to be considered more: 
- there is no 'lock' on the wallets directory, which might be needed?
- because this uses a new wallets/ directory by default, downgrading to an earlier version won't see the wallets in that directory (not a big deal though, users can just copy them up to the main dir)
- jnewbery suggested putting each wallet in its own directory, which is a good idea, but out of scope for this PR IMO. EDIT: this is being done in https://github.com/bitcoin/bitcoin/pull/11687
- doc/files.md needs updating (will do soon)

I also considered including  a cleanup by removing caching of data directory paths and instead just initialise them once on startup (c.f. #3073), but decided it wasn't super relevant here will just complicate review.

meshcollider committed:
Add -walletdir parameter to specify custom wallet dir

meshcollider committed:
Add test for -walletdir

meshcollider committed:
Add release notes for -walletdir and wallets/ dir

promag commented on src/util.cpp:
```diff
@@ -590,6 +590,10 @@ fs::path GetWalletDir()
     } else {
         path = GetDataDir();
+        // If a wallets directory exists, use that, otherwise default to GetDataDir
+        if (fs::is_directory(path / "wallets")) {
```
In commit *Default wallet directory is wallets/ if it exists*, missing test?

MeshCollider replied:
This is part of the WIP, I'll add a test for that when I make it actually initialise the directory :)

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2626,17 +2626,16 @@ UniValue listwallets(const JSONRPCRequest& request)
-            "  \"walletname\"            (string) the wallet name\n"
-            "   ...\n"
-            "]\n"
+            "{\n"
```
This is breaking change. Also, this is not so relevant to this PR? I suggest to create a PR with this to separate discussions.

promag replied:
Another interface is:
```json
{
  "wallets": [
    { "name": "foo", "loaded": true },
    { "name": "bar", "loaded": false }
  ]
}
```
Which has the advantage of allowing to add new attributes later.

MeshCollider replied:
Re: breaking change, I believe it was only added in 0.15 anyway so its expected to be more unstable. Can separate into a new PR if desired

MeshCollider replied:
Also re: interface, that idea is nice but it doesn't work well with the current logic of looking for `*.dat` files imo, `available` is slightly less suggestive that they *are* actually wallets than putting them all in a `wallets` list, you'd end up with `{ "name": "peers.dat", "loaded": false}` if you were using the datadir. If you or anyone has a better idea for finding wallets then that would be ok

ryanofsky replied:
> If you or anyone has a better idea for finding wallets then that would be ok

You could look for berkeleydb magic bytes to at least filter out files which can't be bdb databases:

https://github.com/file/file/blob/master/magic/Magdir/database

promag replied:
What is the goal? Why anyone wants to list available wallets? I guess that whoever wants multiwallet keeps a record of the wallets externally? Anyway, I strongly suggest to move that discussion to a new PR.

jnewbery replied:
This change may have been influenced by my comment here: https://github.com/bitcoin/bitcoin/issues/11348#issuecomment-330234723. See the original post in that issue for motivation.

My view is that `listwallets` was announced as experimental in v0.15 (along with all the other multiwallet RPCs) so we have a bit more licence to make breaking changes than normal. I think this change should stay, but like @MeshCollider says, there may be better ways to achieve it.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2645,10 +2644,23 @@ UniValue listwallets(const JSONRPCRequest& request)
-    return obj;
+    UniValue available(UniValue::VARR);
+    fs::path walletdir = GetWalletDir();
+    for (fs::directory_iterator it(walletdir); it != fs::directory_iterator(); it++) {
```
`++it`.

promag commented on src/util.cpp:
```diff
@@ -579,6 +579,26 @@ const fs::path &GetDataDir(bool fNetSpecific)
+    } else {
+        path = GetDataDir();
+        // If a wallets directory exists, use that, otherwise default to GetDataDir
+        if (fs::is_directory(path / "wallets")) {
```
When is this directory created?

MeshCollider replied:
It's not yet, it's on the to-do list ;)

promag commented on src/util.cpp:
```diff
@@ -579,6 +579,26 @@ const fs::path &GetDataDir(bool fNetSpecific)
+    if (gArgs.IsArgSet("-walletdir")) {
+        path = fs::system_complete(gArgs.GetArg("-walletdir", ""));
+        if (!fs::is_directory(path)) {
+            path = "";
```
At the moment `GetWalletDir` is used in the startup and in runtime.

IMO, in runtime, this should not be possible. Create a `bool IsValidWalletDir()` (or whatever) to use in startup and here assert `is_directory`?

MeshCollider replied:
Yep I'll include something similar when reworking the directory caching stuff I think, datadir and walletdir shouldn't change throughout execution so no point rechecking every time

promag commented on src/qt/bitcoin.cpp:
```diff
@@ -626,6 +626,12 @@ int main(int argc, char *argv[])
                               QObject::tr("Error: Specified data directory \"%1\" does not exist.").arg(QString::fromStdString(gArgs.GetArg("-datadir", ""))));
         return EXIT_FAILURE;
     }
+    if (gArgs.IsArgSet("-walletdir") && !fs::is_directory(GetWalletDir()))
```
`{` in this line.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2645,10 +2644,23 @@ UniValue listwallets(const JSONRPCRequest& request)
+    for (fs::directory_iterator it(walletdir); it != fs::directory_iterator(); it++) {
+        if (is_regular_file(*it)
+         && it->path().filename().string().length() > 4
+         && it->path().filename().string().substr(it->path().filename().string().length() - 4) == ".dat") {
```
This is kind of hard to read. Care to simplify?

promag commented on doc/release-notes.md:
```diff
@@ -74,6 +74,13 @@ Due to a backward-incompatible change in the wallet database, wallets created
 with version 0.16.0 will be rejected by previous versions. Also, version 0.16.0
 will only create hierarchical deterministic (HD) wallets.
 
+Custom wallet directories
```
👏 to release notes.

promag commented on doc/release-notes.md:
```diff
@@ -74,6 +74,13 @@ Due to a backward-incompatible change in the wallet database, wallets created
+---------------------
+The ability to specify a directory other than the default data directory in which to store
+wallets has been added. An existing directory can be specified using the `-walletdir=<dir>`
+argument. The parameter defaults to using the data directory location if not included.
```
Missing `datadir/wallets` case?

MeshCollider replied:
Will add when implemented :+1:

promag commented on src/init.cpp:
```diff
@@ -367,6 +367,7 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-sysperms", _("Create new files with system default permissions, instead of umask 077 (only effective with disabled wallet functionality)"));
 #endif
     strUsage += HelpMessageOpt("-txindex", strprintf(_("Maintain a full transaction index, used by the getrawtransaction rpc call (default: %u)"), DEFAULT_TXINDEX));
+    strUsage += HelpMessageOpt("-walletdir=<dir>", _("Specify wallet directory"));
```
Specify directory to hold wallets?

promag commented on doc/release-notes.md:
```diff
@@ -74,6 +74,13 @@ Due to a backward-incompatible change in the wallet database, wallets created
+Custom wallet directories
+---------------------
+The ability to specify a directory other than the default data directory in which to store
+wallets has been added. An existing directory can be specified using the `-walletdir=<dir>`
```
Nit, should be `-walletsdir` (plural)?

MeshCollider replied:
I prefer walletdir, no real difference though imo

laanwj replied:
`walletsdir` seems like something I'd fat finger into `walletdir` every time, prefer to leave out the extra letter :)

promag reviewed (commented):
Concept ACK. Regardless of being WIP, some comments.

jonasschnelli commented:
I haven't looked at the code, but how is this interacting with the multi-wallet-mode?
Does it allow a directory per wallet?

MeshCollider commented:
@jonasschnelli no, all wallets would be inside the walletdir specified

MeshCollider commented:
Addressed @promag nits, will work on the directory creation stuff as soon as I can

jnewbery commented on src/init.cpp:
```diff
@@ -367,6 +367,7 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-sysperms", _("Create new files with system default permissions, instead of umask 077 (only effective with disabled wallet functionality)"));
 #endif
     strUsage += HelpMessageOpt("-txindex", strprintf(_("Maintain a full transaction index, used by the getrawtransaction rpc call (default: %u)"), DEFAULT_TXINDEX));
+    strUsage += HelpMessageOpt("-walletdir=<dir>", _("Specify directory to hold wallets"));
```
I think this help text should go in `GetWalletHelpString()` in `src/wallet/init.cpp`

jnewbery commented on src/init.cpp:
```diff
@@ -1224,6 +1225,7 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
         LogPrintf("Startup time: %s\n", DateTimeStrFormat("%Y-%m-%d %H:%M:%S", GetTime()));
     LogPrintf("Default data directory %s\n", GetDefaultDataDir().string());
     LogPrintf("Using data directory %s\n", GetDataDir().string());
+    LogPrintf("Using wallet directory %s\n", GetWalletDir().string());
```
Again, I think this log should go in the wallet module rather than server.

jnewbery commented on src/util.cpp:
```diff
@@ -579,6 +579,26 @@ const fs::path &GetDataDir(bool fNetSpecific)
+    if (gArgs.IsArgSet("-walletdir")) {
+        path = fs::system_complete(gArgs.GetArg("-walletdir", ""));
+        if (!fs::is_directory(path)) {
+            path = "";
```
I don't love this. At the moment you're using `path = ""` to essentially return an error to the caller (since on start-up, the calling function checks that the returned `fs::path` is a directory). That's not at all clear from reading just this function.

Can you be more explicit about how the error is returned, either by explicitly returning bool or at the least by adding a comment here?

jnewbery commented on src/bitcoin-cli.cpp:
```diff
@@ -105,6 +105,10 @@ static int AppInitRPC(int argc, char* argv[])
         fprintf(stderr, "Error: Specified data directory \"%s\" does not exist.\n", gArgs.GetArg("-datadir", "").c_str());
         return EXIT_FAILURE;
     }
+    if (gArgs.IsArgSet("-walletdir") && !fs::is_directory(GetWalletDir())) {
```
Is this necessary? Where is the wallet name used in `bitcoin-cli.cpp`?

jnewbery commented on test/functional/multiwallet.py:
```diff
@@ -15,30 +15,41 @@ class MultiWalletTest(BitcoinTestFramework):
         self.setup_clean_chain = True
         self.num_nodes = 1
-        self.extra_args = [['-wallet=w1', '-wallet=w2', '-wallet=w3']]
+        self.extra_args = [['-wallet=w1.dat', '-wallet=w2.dat', '-wallet=w3.dat']]
```
Wallet names are currently allowed to not end with `.dat`. This test verified that we don't regress that support.

I'm not sure if we should go on supporting that, but we should discuss that before removing support.

MeshCollider replied:
Agreed, I'm unsure if there are any real benefits of adding an `available` list to the listwallets RPC after all, or if it just complicates the PR. Will discuss on IRC and then remove if not.

jnewbery commented on test/functional/multiwallet.py:
```diff
@@ -15,30 +15,41 @@ class MultiWalletTest(BitcoinTestFramework):
+        # running the node with specified walletdir should only have the default wallet in it
+        os.mkdir(os.path.join(self.options.tmpdir, 'node0', 'regtest', 'walletdir'))
+        self.start_node(0, ['-walletdir=' + os.path.join(self.options.tmpdir, 'node0', 'regtest', 'walletdir')])
+        assert_equal(set(self.nodes[0].listwallets()['loaded']), {"wallet.dat"})
```
This test should also test having multiple wallets in the walletdir, and issuing an RPC command to one of them.

jnewbery commented on doc/release-notes.md:
```diff
@@ -74,6 +74,13 @@ Due to a backward-incompatible change in the wallet database, wallets created
+The ability to specify a directory other than the default data directory in which to store
+wallets has been added. An existing directory can be specified using the `-walletdir=<dir>`
+argument. The parameter defaults to using the data directory location if not included.
+Wallets loaded via `-wallet` arguments are relative to this wallet directory.
```
`relative to this wallet directory` or `must be in this wallet directory`

jnewbery replied:
I'm not sure, but perhaps we should add a warning that `-walletdir` must refer to a reliable directory in the filesystem. For example, I think we want to discourage people using ephemeral, removable storage for their `-walletdir` since if the directory becomes unavailable during operation, funds may be lost.

jnewbery commented on src/wallet/init.cpp:
```diff
@@ -208,7 +208,7 @@ bool VerifyWallets()
 
         std::string strError;
-        if (!CWalletDB::VerifyEnvironment(walletFile, GetDataDir().string(), strError)) {
+        if (!CWalletDB::VerifyEnvironment(walletFile, GetWalletDir().string(), strError)) {
```
The arguments in the declaration and definition for `VerifyEnvironment()` and `VerifyDatabaseFile()` are still `dataDir`. Those should be updated to indicate that they're now refering to the wallet directory.

MeshCollider replied:
Fixed

jnewbery reviewed (commented):
This looks like a good start. I've add a few comments inline. Some more high-level feedback:

- The `GetWalletDir()` is currently in utils.cpp, which is part of the libbitcoin_util module. I think it makes more sense to move it to libbitcoin_wallet, since none of the other modules (server, qt, etc) should need to know about it.
- On startup, I think that if this is the first run *AND* there is no `-walletdir`, then bitcoind should create the `wallet` dir in the datadir (ie fresh installs should default to using the `datadir/wallet` structure).

MeshCollider commented:
I've split the `listwallets` RPC changes out, will open a new PR for them. Squashed nit-fix commits, reworked the datadir caching stuff. Needs some careful testing and review but its almost ready to remove WIP.

jnewbery commented:
It might eventually make sense to have each wallet in each own directory, ie the structure would be:

```
/<dataDir>
  .
  .-/wallets
      .
      .-/wallet1
      .   .
      .   .-wallet1.dat
      .   .-db.log
      .   ./database
      .
      .-/wallet2
      .   ....
      ....
 ```

That way, every wallet has its own bdb environment and we avoid nasty surprises like https://github.com/bitcoin/bitcoin/issues/11429#issuecomment-335571653.

That might involve a large code change, which would be necessary anyway if we want to separate wallets into their own processes.

I don't think that has any impact on this PR right now, but it's worth considering what we want the final structure to look like.

promag commented:
Since we are relying on filesystem then :+1: to @jnewbery suggestion.

meshcollider committed:
Create walletdir if datadir doesn't exist and fix tests

meshcollider committed:
Make debugging test crash easier

MeshCollider commented:
Rebased

jnewbery commented on test/functional/multiwallet.py:
```diff
@@ -26,13 +26,34 @@ def run_test(self):
 
         # should not initialize if wallet file is a directory
-        os.mkdir(os.path.join(self.options.tmpdir, 'node0', 'regtest', 'w11'))
+        os.mkdir(os.path.join(self.options.tmpdir, 'node0', 'regtest', 'wallets', 'w11'))
```
Perhaps make some local variables here rather than repeating the same calls to `os.path.join()`:

```python
wallet_dir = os.path.join(self.options.tmpdir, 'node0', 'regtest', 'wallets')
...
wallet_dir2 = os.path.join(self.options.tmpdir, 'node0', 'regtest', 'walletdir')
```

or whatever

jnewbery commented on src/wallet/walletutil.cpp:
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
```
Are you Satoshi? :slightly_smiling_face: If not, I guess you can remove this copyright line (and the same from `walletutil.h`)

jnewbery reviewed (commented):
Looks good. A few minor nits inline.

This PR could do with some judicious squashing. I'm not a fan of PRs adding and then removing code in later commits. Consider:

- squashing commit 'Move GetWalletDir to new walletutil file' into the initial 'Add -walletdir parameter to specify custom wallet dir' commit
- squashing the two release notes commits together.
- squashing 'Fix tests and test wallets/ subdir behaviour' into the 'Create walletdir if datadir doesn't exist' commit (test/code change atomicity is good because it doesn't break git bisect)

meshcollider committed:
Default walletdir is wallets/ if it exists

MeshCollider commented:
Fixed @jnewbery's nits and squashed as suggested. Thanks :)

MarcoFalke commented:
Concept ACK c36cb54711fe7677db6efc4db00a6e7d42e62f8d

ryanofsky commented:
Adding a -walletdir option seems more cumbersome than just tweaking the existing -wallet option to allow directory separator characters (\\/).

Advantages of allowing -wallet directory paths instead of adding -walletdir option:
- Simpler for users: they don't need to look at 3 different options (-wallet, -walletdir, -datadir) to figure out where wallet will be stored. They just need to look at 1 or 2 (-wallet and -datadir if wallet path is not absolute).
- More flexible: Users can store wallets in different locations.
- Doesn't add 100 lines of code.
- 100% backwards compatible. New bitcoin releases continue to store wallet.dat in same places releases.

Disadvantages of allowing -wallet directory paths instead of adding -walletdir option:
- Because users are not forced to store all wallets in single directory, we lose the ability to have a listwallets RPC that can return a complete list of all possible wallets (though it's still possible to list wallets in datadir).

Is this a fair analysis or am I missing something?

sipa commented:
Supporting multiple wallet directories also requires support for multiple BDB environments first, which would be extra code complication.

ryanofsky commented:
> Supporting multiple wallet directories also requires support for multiple BDB environments first, which would be extra code complication.

Oh right, that is the thing I forgot. Allowing -wallet paths would be a little more complicated to implement than I was suggesting, but not much more. And this by itself doesn't seem like a good enough reason to make user experience worse and add complexity that we will need to support into the future.

MeshCollider commented:
> @ryanofsky 100% backwards compatible. New bitcoin releases continue to store wallet.dat in same places releases.

Note that this PR doesn't rely on moving the default directory location, that was just suggested as a good idea in https://github.com/bitcoin/bitcoin/issues/11348 (that was the main point of that issue). So that point is orthogonal to `-walletdir` or a relative `-wallet`, and probably should still be done

MarcoFalke commented:
> ... or a relative -wallet, and probably should still be done

I don't see how we could reasonably test behavior when `-datadir`, `-walletdir` and `-wallet` support directories as paths. Extending `-wallet` to accept file names, relative paths (to the datadir) or absolute paths should make the need to specify a `-walletdir` obsolete.

MeshCollider commented:
@MarcoFalke sorry I meant the moving of default walletdir to datadir/wallets/ not making it accept relative paths :)

MarcoFalke commented:
Yeah, right. Why would you change the default wallet dir when already could specify a path through `-wallet`?

MeshCollider commented:
Just to keep the wallets and their BDB environments contained in their own subdirectory, a cleanup and bit more separation. Would make it clearer to users what they should back up, for example, in the case of multiwallet. That's what #11348 was discussing

sipa commented:
An alternative idea would be to not have a `-walletdir` option, but have it be implied from the `-wallet` arguments. In a first version, a requirement could exist that all wallet files are in the same directory (which would then become the db env directory).

luke-jr commented:
IMO it's a bad idea to infer the wallet dir from the wallet filename. This would encourage confusion among users with regard to the db.log file and database directory - they'll wonder why they exist, possibly delete them, and/or not notice and neglect to back them up (although backups ought to always be made using Core, some users still don't know that).

luke-jr commented on src/util.cpp:
```diff
@@ -574,7 +574,10 @@ const fs::path &GetDataDir(bool fNetSpecific)
+    if (fs::create_directories(path)) {
+        // This is the first run, create wallets subdirectory too
+        fs::create_directories(path / "wallets");
+    }
```
Here and above: We should never create the wallets directory if wallet support is disabled. Additionally, if we're going to default to a wallet subdir, we should create it when there is no wallet yet, regardless of whether the datadir exists or not (eg, if datadir exists, but wallet.dat does not, make and use the wallet subdir).

jnewbery replied:
I had the same reaction when I first saw this, but I don't think it's a problem. If wallet support is disabled, we'll just have an empty wallet directory in the datadir.

laanwj replied:
I think it's fine to create an empty wallets directory, no matter if wallets are enabled or not (either at build or runtime). Aesthetically it would be better otherwise but practically it makes sense:

As the presence of this directory is used as signal where to put wallets, transitioning to it for an existing data directory is non-trivial. How would it even know if a legacy datadir has no wallets and you're creating the first one? That would be some nightmare heuristic logic...

luke-jr commented:
Perhaps another option would be to allow specifying a directory (ending in a `/`) for `-wallet`, and treat the entire directory as dedicated to that wallet, including naming the file itself `wallet.dat` unconditionally...?

laanwj commented:
I think we should go forward with this, instead of turning this into a big argument. `walletdir` might not be the perfect solution for everything (e.g. it cannot handle multiple wallets in different places), but for people storing their wallets on a different filesystem this is enough, and much safer than the dangerous symlink tricks they're using currently.

MeshCollider commented:
Rebased

laanwj commented on src/qt/intro.cpp:
```diff
@@ -214,7 +214,10 @@ bool Intro::pickDataDirectory()
             dataDir = intro.getDataDirectory();
             try {
-                TryCreateDirectories(GUIUtil::qstringToBoostPath(dataDir));
+                if(TryCreateDirectories(GUIUtil::qstringToBoostPath(dataDir))) {
```
u-nit: space after `if`

MeshCollider commented:
@laanwj nit addressed, and travis is passing, so shall I squash the last two commits now?

laanwj commented on src/wallet/walletutil.cpp:
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2017 The Bitcoin Core developers
```
Agree on creating a separate walletutil, this avoids picking up #ifdef WALLET_ENABLE logic in util.cpp.

laanwj commented on src/wallet/init.cpp:
```diff
@@ -34,6 +35,7 @@ std::string GetWalletHelpString(bool showDebug)
     strUsage += HelpMessageOpt("-upgradewallet", _("Upgrade wallet to latest format on startup"));
     strUsage += HelpMessageOpt("-wallet=<file>", _("Specify wallet file (within data directory)") + " " + strprintf(_("(default: %s)"), DEFAULT_WALLET_DAT));
     strUsage += HelpMessageOpt("-walletbroadcast", _("Make the wallet broadcast transactions") + " " + strprintf(_("(default: %u)"), DEFAULT_WALLETBROADCAST));
+    strUsage += HelpMessageOpt("-walletdir=<dir>", _("Specify directory to hold wallets"));
```
Need information about the default here: `<datadir>/wallets if it exists, otherwise <datadir>`

laanwj commented on src/wallet/init.cpp:
```diff
@@ -230,7 +238,7 @@ bool VerifyWallets()
 
         std::string strWarning;
-        bool dbV = CWalletDB::VerifyDatabaseFile(walletFile, GetDataDir().string(), strWarning, strError);
+        bool dbV = CWalletDB::VerifyDatabaseFile(walletFile, GetWalletDir().string(), strWarning, strError);
```
Seems overkill to call GetWalletDir() so many times in this function, we could just call and store it once. But no, no need to fix in this PR, this is easier to review.

laanwj commented on src/wallet/db.cpp:
```diff
@@ -282,18 +283,18 @@ bool CDB::VerifyEnvironment(const std::string& walletFile, const fs::path& dataD
+        if (!bitdb.Open(walletDir)) {
             // if it still fails, it probably means we can't even create the database env
-            errorStr = strprintf(_("Error initializing wallet database environment %s!"), GetDataDir());
+            errorStr = strprintf(_("Error initializing wallet database environment %s!"), GetWalletDir());
```
You already have walletDir here, just log that?

laanwj commented on doc/release-notes.md:
```diff
@@ -62,6 +62,19 @@ Due to a backward-incompatible change in the wallet database, wallets created
+when choosing a wallet directory location, as if  it becomes unavailable during operation,
+funds may be lost.
+
+On new installations (if the data directory doesn't exist), this parameter will default to a
```
I think this is a very important change from the viewpoint of users ("where's my wallet.dat?! all wiki articles say it's in the data directory!"), and probably warrants its own section.

MeshCollider replied:
Agreed, although I would hope most users would work out what the new "wallets/" directory was for ;)

laanwj commented:
Manually tested:
- Current datadir (testnet) - ok, still uses existing wallet
- Create new datadir (testnet) - ok, `wallets` is created and wallet.dat is there
- Create new datadir (testnet), different `-wallet=test.dat` name - ok, `wallets` is created and `test.dat` is there
- `-choosedatadir` with GUI, new datadir - ok, `wallets` is created and wallet.dat is there
- `-choosedatadir` with GUI, re-select default datadir - ok, datadir is left as-is, uses existing wallet

MeshCollider commented:
Squashed fixup commits

laanwj commented:
Tested ACK c1e5d40

promag commented on src/qt/intro.cpp:
```diff
@@ -214,7 +214,10 @@ bool Intro::pickDataDirectory()
-                TryCreateDirectories(GUIUtil::qstringToBoostPath(dataDir));
+                if (TryCreateDirectories(GUIUtil::qstringToBoostPath(dataDir))) {
+                    // If a new data directory has been created, make wallets subdirectory too
+                    TryCreateDirectories(GUIUtil::qstringToBoostPath(dataDir) / "wallets");
```
The logic data_dir + "/wallets" exists in a couple of places. Doesn't this also ignores -walletdir?

laanwj replied:
This is just for creating the initial wallets directory. If it's not used it will stay empty.

promag commented on src/wallet/walletutil.cpp:
```diff
@@ -0,0 +1,27 @@
+        path = fs::system_complete(gArgs.GetArg("-walletdir", ""));
+        if (!fs::is_directory(path)) {
+            // If the path specified doesn't exist, we return the deliberately
+            // invalid empty string.
```
Why? Couldn't create it? Is it bad setting?

laanwj replied:
Unlike the datadir, `walletdir` is not created if it doesn't exist. This throws an error in init:
```
$ src/bitcoind -regtest -walletdir=/tmp/walletsdf
Error: Error: Specified wallet directory "/tmp/walletsdf" does not exist.
```
That's the right behavior, IMO.

promag replied:
Nit, if `/tmp/walletsdf` is non-directory then the error could be different.

laanwj commented:
There's a small thing I'm unsure about here, which just came to me, which only affects testnet and regtest: do we want to subdir the `-walletdir` with the chain name? I think this would make sense, because currently it's impossible to specify `walletdir` in bitcoin.conf without running into conflicts.

promag commented:
Tested ACK c1e5d40.

Nits and improvements to regtest and testnet can be left for followups?

ryanofsky commented:
You guys should look at #11687, too. It works standalone, but combined with this PR, it basically implements what john suggested here: https://github.com/bitcoin/bitcoin/pull/11466#issuecomment-335827526

It also has a number of other improvements. I'm currently fixing a crazy problem with the tests (they were only failing on travis and I spent hours yesterday trying to reproduce them locally before eventually succeeding this morning and finding a shutdown bug.) Luke-jr posted a very reasonable comment there objecting to an earlier version of the PR that didn't include the third commit, but I think that should be addressed now. I'm planning on posting an update there after I fix travis, but the PR is very well documented and includes detailed release notes that I think should be pretty clear.

laanwj commented:
@ryanofsky will look at that one next

> Nits and improvements to regtest and testnet can be left for followups?

Yes, fine with me, will leave it up to @MeshCollider .

MeshCollider commented:
> Yes, fine with me, will leave it up to @MeshCollider .

Yep I'm happy to open a follow-up PR

jonasschnelli commented on doc/release-notes.md:
```diff
@@ -62,6 +62,20 @@ Due to a backward-incompatible change in the wallet database, wallets created
+The ability to specify a directory other than the default data directory in which to store
+wallets has been added. An existing directory can be specified using the `-walletdir=<dir>`
+argument. Wallets loaded via `-wallet` arguments must be in this wallet directory. Care should be taken
+when choosing a wallet directory location, as if  it becomes unavailable during operation,
```
nit: double space (`if  it`).

MeshCollider commented:
Nit cleanup and regtest/testnet change on new branch here: https://github.com/MeshCollider/bitcoin/tree/201711_walletdir
The second commit requires more testing so best if I leave them both for a new PR, approach can be discussed there too
